docs/references/security/rust-canister-development-security-best-practices.md

```diff
@@ -183,43 +183,159 @@ A canister could be rendered unusable so it could never be upgraded again e.g. d
 
 This may e.g. lead to the following issues:
 
--   If state changes before an inter-canister call leave the state inconsistent and there is a panic after the inter-canister call, this results in inconsistent canister state.
+- If state changes before an inter-canister call leave the state inconsistent and there is a panic after the inter-canister call, this results in inconsistent canister state.
 
--   In particular, if allocated resources (e.g. locks or memory) from before an inter-canister call are not released this can e.g. lead to a canister being locked forever.
+- In particular, if e.g. part of the canister state is locked before an inter-canister call and released in the callback, the lock may never be released if the callback traps. 
 
--   Generally, there can be bugs when data is not persisted when the developer expected it to be.
+- Generally, there can be bugs when data is not persisted when the developer expected it to be.
+
+Note that in Rust, from Rust CDK version 0.5.1, any local variables still go out of scope on panic. This prevents some of the issues above. 
 
 #### Recommendation
 
--   [Don’t panic after `await`](https://mmapped.blog/posts/01-effective-rust-canisters.html#panic-await) (from [Effective Rust Canisters](https://mmapped.blog/posts/01-effective-rust-canisters.html))
+- Watch the [community conversation on Security Best Practices](https://www.youtube.com/watch?v=PneRzDmf_Xw&list=PLuhDt1vhGcrez-f3I0_hvbwGZHZzkZ7Ng&index=2&t=4s) which shows a concrete example of an issue as described here.
+
+- [Don’t lock shared resources across await boundaries](https://mmapped.blog/posts/01-effective-rust-canisters.html#dont-lock) (from [Effective Rust Canisters](https://mmapped.blog/posts/01-effective-rust-canisters.html))
 
--   [Don’t lock shared resources across await boundaries](https://mmapped.blog/posts/01-effective-rust-canisters.html#dont-lock) (from [Effective Rust Canisters](https://mmapped.blog/posts/01-effective-rust-canisters.html))
+- [Don’t panic after `await`](https://mmapped.blog/posts/01-effective-rust-canisters.html#panic-await) (from [Effective Rust Canisters](https://mmapped.blog/posts/01-effective-rust-canisters.html))
 
--   See also: "Inter-canister calls" section in [How to audit an Internet Computer canister](https://www.joachim-breitner.de/blog/788-How_to_audit_an_Internet_Computer_canister)
+- See also: "Inter-canister calls" section in [How to audit an Internet Computer canister](https://www.joachim-breitner.de/blog/788-How_to_audit_an_Internet_Computer_canister)
 
--   For context: [IC interface spec on message execution](../ic-interface-spec#message-execution)
+- For context: [IC interface spec on message execution](../ic-interface-spec#message-execution)
 
-### Be aware that state may change during inter-canister calls
+### Be aware that there is no reliable message ordering
```

Comment:
How about also raising awareness that there is no reliable messaging? The IC can fail any inter-canister call even if both the sender and the receiver are up and have sufficient cycles (because IC-internal data structures get full or whatever). So the canisters always have to handle the failure of such a call.

Reply:
Thanks! I like the text you wrote, but I think we should also state this as a basic property of our messaging, since, well, it is a property of our messaging :) I added a couple of suggestions

Reply:
That makes sense. Added your suggestions. Thanks!